### ISSUE TITLE
[bug]: Translation path for G729 not enabled at all

### ISSUE THREAD

**Issue by guayomejiae** [FreePBX/issue-tracker#578]
### FreePBX Version

FreePBX 17

### Issue Description

FreePBX v17 GA installed using the script on Debian 12 OS. 
All setup worked as expected, gui and asterisk are working fine; the only thing isn´t working is "translation" for the g.729 codec > alaw. 

Extension 1099 running on g.729 codec makes a call through a trunk supporting only alaw, the call ends and verbose shows: 

`set_format: Unable to find a codec translation path: (alaw) -> (g729)`
`set_format: Unable to find a codec translation path: (g729) -> (alaw)`

Running the command "core show translation paths g729" shows:

`g729:8000        To ulaw:8000       : No Translation Path`
`g729:8000        To alaw:8000       : No Translation Path`

Checking the modules loaded, I have "codec_resample.so" already loaded. 

`codec_resample.so              SLIN Resampling Codec                    0          Running              core`

### Operating Environment

FreePBX 17.0.19.21
Asterisk 20.10.0 built by jenkins @ fdd18907e9c7 on a x86_64 running Linux >>> This version is needed because LTS (v21 isn´t LTS)
Debian GNU/Linux 12 (bookworm)

### Relevant log output

```shell
-- PJSIP/1099-0000000a Internal Gosub(func-apply-sipheaders,s,1) complete GOSUB_RETVAL=
    -- Called PJSIP/1099/sip:1099@172.16.111.103:60955;ob
  == Using SIP RTP Audio TOS bits 184
  == Using SIP RTP Audio TOS bits 184 in TCLASS field.
  == Using SIP RTP Audio CoS mark 5
    -- PJSIP/1099-0000000a is ringing
[2024-12-05 12:26:59] WARNING[1616]: channel.c:5791 set_format: Unable to find a codec translation path: (g729) -> (alaw)
[2024-12-05 12:26:59] WARNING[1616]: channel.c:5791 set_format: Unable to find a codec translation path: (alaw) -> (g729)
    -- PJSIP/1099-0000000a answered PJSIP/PBX_office-00000009
[2024-12-05 12:26:59] WARNING[7602][C-0000000a]: channel.c:6782 ast_channel_make_compatible_helper: No path to translate from PJSIP/1099-0000000a to PJSIP/PBX_office-00000009
[2024-12-05 12:26:59] WARNING[7602][C-0000000a]: app_dial.c:3503 dial_exec_full: Had to drop call because I couldn't make PJSIP/PBX_office-00000009 compatible with PJSIP/1099-0000000a
    -- PJSIP/1099-0000000a Internal Gosub(app-missedcall-hangup,1099,1) start
    -- Executing [1099@app-missedcall-hangup:1] NoOp("PJSIP/1099-0000000a", "Dialed: 1099") in new stack
    -- Executing [1099@app-missedcall-hangup:2] NoOp("PJSIP/1099-0000000a", "Caller: ") in new stack
    -- Executing [1099@app-missedcall-hangup:3] GotoIf("PJSIP/1099-0000000a", "0?exit") in new stack
    -- Executing [1099@app-missedcall-hangup:4] Set("PJSIP/1099-0000000a", "EXTENNUM=1099") in new stack
    -- Executing [1099@app-missedcall-hangup:5] Set("PJSIP/1099-0000000a", "FEXTENNUM=1099") in new stack
    -- Executing [1099@app-missedcall-hangup:6] GotoIf("PJSIP/1099-0000000a", "0?exit") in new stack
    -- Executing [1099@app-missedcall-hangup:7] AGI("PJSIP/1099-0000000a", "agi://127.0.0.1/missedcallnotify.php,1099,,1099,0,,PJSIP/1099-0000000a,,,,") in new stack
    -- <PJSIP/1099-0000000a>AGI Script agi://127.0.0.1/missedcallnotify.php completed, returning 0
    -- Executing [1099@app-missedcall-hangup:8] Return("PJSIP/1099-0000000a", "") in new stack
  == Spawn extension (from-internal, , 1) exited non-zero on 'PJSIP/1099-0000000a'
    -- PJSIP/1099-0000000a Internal Gosub(app-missedcall-hangup,1099,1) complete GOSUB_RETVAL=
  == Spawn extension (dialOne-with-exten, 1099, 2) exited non-zero on 'PJSIP/PBX_office-00000009'
    -- PJSIP/PBX_office-00000009 Internal Gosub(app-missedcall-hangup,1099,1) start
    -- Executing [1099@app-missedcall-hangup:1] NoOp("PJSIP/PBX_office-00000009", "Dialed: 1099") in new stack
    -- Executing [1099@app-missedcall-hangup:2] NoOp("PJSIP/PBX_office-00000009", "Caller: ") in new stack
    -- Executing [1099@app-missedcall-hangup:3] GotoIf("PJSIP/PBX_office-00000009", "0?exit") in new stack
    -- Executing [1099@app-missedcall-hangup:4] Set("PJSIP/PBX_office-00000009", "EXTENNUM=1099") in new stack
    -- Executing [1099@app-missedcall-hangup:5] Set("PJSIP/PBX_office-00000009", "FEXTENNUM=1099") in new stack
    -- Executing [1099@app-missedcall-hangup:6] GotoIf("PJSIP/PBX_office-00000009", "0?exit") in new stack
    -- Executing [1099@app-missedcall-hangup:7] AGI("PJSIP/PBX_office-00000009", "agi://127.0.0.1/missedcallnotify.php,1099,,1099,0,,PJSIP/PBX_office-00000009,ANSWER,,,") in new stack
    -- <PJSIP/PBX_office-00000009>AGI Script agi://127.0.0.1/missedcallnotify.php completed, returning 0
    -- Executing [1099@app-missedcall-hangup:8] Return("PJSIP/PBX_office-00000009", "") in new stack
  == Spawn extension (dialOne-with-exten, 1099, 2) exited non-zero on 'PJSIP/PBX_office-00000009'
```

**Comment by jcolp** [comment]
There is codec module included for the translation of g729. The codec_resample is a resampling module for going between different bitrates in signed linear.

**Comment by guayomejiae** [comment]
Thanks Joshua, codec g729 is already installed and working fine, it is just
the translation path missing:

[image: imagen.png]

Codec is loaded:

[image: imagen.png]

I think the asterisk compilation script isn´t including something related
to translation paths between those codecs....
Ed.

**Comment by guayomejiae** [comment]
->>> **UPDATED ENVIORNMET for testing purposes** <<<-

### Operating Environment

FreePBX 17.0.19.21
Asterisk 21.5.0 
Debian GNU/Linux 12 (bookworm)

### Relevant log output

On asterisk CLI: 

`core show translation paths g729`

Result: 

> --- Translation paths SRC Codec "g729" sample rate 8000 ---
        g729:8000        To codec2:8000     : No Translation Path
        g729:8000        To g723:8000       : No Translation Path
        g729:8000        To ulaw:8000       : No Translation Path
        g729:8000        To alaw:8000       : No Translation Path
        g729:8000        To gsm:8000        : No Translation Path
        g729:8000        To g726:8000       : No Translation Path
        g729:8000        To g726aal2:8000   : No Translation Path
        g729:8000        To adpcm:8000      : No Translation Path
        g729:8000        To slin:8000       : No Translation Path
        g729:8000        To slin:12000      : No Translation Path
        g729:8000        To slin:16000      : No Translation Path
        g729:8000        To slin:24000      : No Translation Path
        g729:8000        To slin:32000      : No Translation Path
        g729:8000        To slin:44100      : No Translation Path
        g729:8000        To slin:48000      : No Translation Path
        g729:8000        To slin:96000      : No Translation Path
        g729:8000        To slin:192000     : No Translation Path
        g729:8000        To lpc10:8000      : No Translation Path
        g729:8000        To speex:8000      : No Translation Path
        g729:8000        To speex:16000     : No Translation Path
        g729:8000        To speex:32000     : No Translation Path
        g729:8000        To ilbc:8000       : No Translation Path
        g729:8000        To g722:16000      : No Translation Path
        g729:8000        To siren7:16000    : No Translation Path
        g729:8000        To siren14:32000   : No Translation Path
        g729:8000        To g719:48000      : No Translation Path
        g729:8000        To opus:48000      : No Translation Path
        g729:8000        To none:8000       : No Translation Path
        g729:8000        To silk:8000       : No Translation Path
        g729:8000        To silk:12000      : No Translation Path
        g729:8000        To silk:16000      : No Translation Path
        g729:8000        To silk:24000      : No Translation Path


Query for alaw codec:

`core show translation paths alaw`

> --- Translation paths SRC Codec "alaw" sample rate 8000 ---
        alaw:8000        To codec2:8000     : (alaw@8000)->(slin@8000)->(codec2@8000)
        **alaw:8000        To g723:8000       : No Translation Path**
        alaw:8000        To ulaw:8000       : (alaw@8000)->(ulaw@8000)
        alaw:8000        To gsm:8000        : (alaw@8000)->(slin@8000)->(gsm@8000)
        alaw:8000        To g726:8000       : (alaw@8000)->(slin@8000)->(g726@8000)
        alaw:8000        To g726aal2:8000   : (alaw@8000)->(slin@8000)->(g726aal2@8000)
        alaw:8000        To adpcm:8000      : (alaw@8000)->(slin@8000)->(adpcm@8000)
        alaw:8000        To slin:8000       : (alaw@8000)->(slin@8000)
        alaw:8000        To slin:12000      : (alaw@8000)->(slin@8000)->(slin@12000)
        alaw:8000        To slin:16000      : (alaw@8000)->(slin@8000)->(slin@16000)
        alaw:8000        To slin:24000      : (alaw@8000)->(slin@8000)->(slin@24000)
        alaw:8000        To slin:32000      : (alaw@8000)->(slin@8000)->(slin@32000)
        alaw:8000        To slin:44100      : (alaw@8000)->(slin@8000)->(slin@44100)
        alaw:8000        To slin:48000      : (alaw@8000)->(slin@8000)->(slin@48000)
        alaw:8000        To slin:96000      : (alaw@8000)->(slin@8000)->(slin@96000)
        alaw:8000        To slin:192000     : (alaw@8000)->(slin@8000)->(slin@192000)
        alaw:8000        To lpc10:8000      : (alaw@8000)->(slin@8000)->(lpc10@8000)
        **alaw:8000        To g729:8000       : No Translation Path**
        alaw:8000        To speex:8000      : (alaw@8000)->(slin@8000)->(speex@8000)
        alaw:8000        To speex:16000     : (alaw@8000)->(slin@8000)->(slin@16000)->(speex@16000)
        alaw:8000        To speex:32000     : (alaw@8000)->(slin@8000)->(slin@32000)->(speex@32000)
        alaw:8000        To ilbc:8000       : (alaw@8000)->(slin@8000)->(ilbc@8000)
        alaw:8000        To g722:16000      : (alaw@8000)->(slin@8000)->(g722@16000)
        alaw:8000        To siren7:16000    : (alaw@8000)->(slin@8000)->(slin@16000)->(siren7@16000)
        alaw:8000        To siren14:32000   : (alaw@8000)->(slin@8000)->(slin@32000)->(siren14@32000)
        alaw:8000        To g719:48000      : No Translation Path
        alaw:8000        To opus:48000      : (alaw@8000)->(slin@8000)->(slin@48000)->(opus@48000)
        **alaw:8000        To none:8000       : No Translation Path**
        alaw:8000        To silk:8000       : (alaw@8000)->(slin@8000)->(silk@8000)
        alaw:8000        To silk:12000      : (alaw@8000)->(slin@8000)->(slin@12000)->(silk@12000)
        alaw:8000        To silk:16000      : (alaw@8000)->(slin@8000)->(slin@16000)->(silk@16000)
        alaw:8000        To silk:24000      : (alaw@8000)->(slin@8000)->(slin@24000)->(silk@24000)

You can see: there is no path for alaw >< g729.

**Comment by blazestudios97** [comment]
Please provide the output for the actual loaded codec modules. The output you are using will show that data regardless if the module is installed or loaded. I get the same exact output and I do not have g729 installed and therefore not loaded to use. If the codec isn't in this list, it isn't loaded for Asterisk to use.

```
hvs01-west*CLI> module show like codec
Module                         Description                              Use Count  Status      Support Level
codec_a_mu.so                  A-law and Mulaw direct Coder/Decoder     0          Running              core
codec_adpcm.so                 Adaptive Differential PCM Coder/Decoder  0          Running              core
codec_alaw.so                  A-law Coder/Decoder                      0          Running              core
codec_codec2.so                Codec 2 Coder/Decoder                    0          Running              core
codec_g722.so                  ITU G.722-64kbps G722 Transcoder         0          Running              core
codec_g726.so                  ITU G.726-32kbps G726 Transcoder         0          Running              core
codec_gsm.so                   GSM Coder/Decoder                        0          Running              core
codec_ilbc.so                  iLBC Coder/Decoder                       0          Running              core
codec_lpc10.so                 LPC10 2.4kbps Coder/Decoder              0          Running              core
codec_opus.so                  OPUS Coder/Decoder                       0          Running          extended
codec_resample.so              SLIN Resampling Codec                    0          Running              core
codec_speex.so                 Speex Coder/Decoder                      1          Running              core
codec_ulaw.so                  mu-Law Coder/Decoder                     0          Running              core
13 modules loaded
```

**Comment by guayomejiae** [comment]
> Please provide the output for the actual loaded codec modules. The output you are using will show that data regardless if the module is installed or loaded. I get the same exact output and I do not have g729 installed and therefore not loaded to use. If the codec isn't in this list, it isn't loaded for Asterisk to use.
> 
> ```
> hvs01-west*CLI> module show like codec
> Module                         Description                              Use Count  Status      Support Level
> codec_a_mu.so                  A-law and Mulaw direct Coder/Decoder     0          Running              core
> codec_adpcm.so                 Adaptive Differential PCM Coder/Decoder  0          Running              core
> codec_alaw.so                  A-law Coder/Decoder                      0          Running              core
> codec_codec2.so                Codec 2 Coder/Decoder                    0          Running              core
> codec_g722.so                  ITU G.722-64kbps G722 Transcoder         0          Running              core
> codec_g726.so                  ITU G.726-32kbps G726 Transcoder         0          Running              core
> codec_gsm.so                   GSM Coder/Decoder                        0          Running              core
> codec_ilbc.so                  iLBC Coder/Decoder                       0          Running              core
> codec_lpc10.so                 LPC10 2.4kbps Coder/Decoder              0          Running              core
> codec_opus.so                  OPUS Coder/Decoder                       0          Running          extended
> codec_resample.so              SLIN Resampling Codec                    0          Running              core
> codec_speex.so                 Speex Coder/Decoder                      1          Running              core
> codec_ulaw.so                  mu-Law Coder/Decoder                     0          Running              core
> 13 modules loaded
> ```

Sure, output is: 

`Module                         Description                              Use Count  Status      Support Level
codec_a_mu.so                  A-law and Mulaw direct Coder/Decoder     0          Running              core
codec_adpcm.so                 Adaptive Differential PCM Coder/Decoder  0          Running              core
codec_alaw.so                  A-law Coder/Decoder                      0          Running              core
codec_ast22_g729.so            Digium G.729 Annex A Codec (optimized fo 0          Running          extended
codec_codec2.so                Codec 2 Coder/Decoder                    0          Running              core
codec_dahdi.so                 Generic DAHDI Transcoder Codec Translato 0          Running              core
codec_g722.so                  ITU G.722-64kbps G722 Transcoder         0          Running              core
codec_g726.so                  ITU G.726-32kbps G726 Transcoder         0          Running              core
codec_gsm.so                   GSM Coder/Decoder                        0          Running              core
codec_ilbc.so                  iLBC Coder/Decoder                       0          Running              core
codec_lpc10.so                 LPC10 2.4kbps Coder/Decoder              0          Running              core
codec_opus.so                  OPUS Coder/Decoder                       0          Running          extended
codec_resample.so              SLIN Resampling Codec                    0          Running              core
codec_silk.so                  Silk Transcoder                          0          Running          extended
codec_siren14.so               ITU G.722.1 Annex C (Siren14, licensed f 0          Running          extended
codec_siren7.so                ITU G.722.1 (Siren7, licensed from Polyc 0          Running          extended
codec_speex.so                 Speex Coder/Decoder                      1          Running              core
codec_ulaw.so                  mu-Law Coder/Decoder                     0          Running              core`
`18 modules loaded
`

**Comment by blazestudios97** [comment]
Do you have a license for the g729? It's required for translations. You can check by doing:

`g729 show licenses`

**Comment by guayomejiae** [comment]
Thanks @blazestudios97 
After analyzing the commands and what you are telling me, I understand now that FreePBX in version 17 will no longer integrate the G.729 codec if a license is not paid for?

Why this new limitation? If in previous versions the codec was always included (because it is already a free-to-use license).

**Comment by blazestudios97** [comment]
From what I understand of the g729 codec, as long as all channels involved are using g729 no license is needed. If you want to translate to/from another codec, you need the license.

So `g729<>g729` works just fine but `g729<>another codec` needs a license.

**Comment by girishmani1992** [comment]
Hi @guayomejiae ,

We have tested in our lab setup and calls are working perfectly with the above mentioned scenario. 

`
core show translation paths g729
--- Translation paths SRC Codec "g729" sample rate 8000 ---
        g729:8000        To codec2:8000     : (g729@8000)->(slin@8000)->(codec2@8000)
        g729:8000        To g723:8000       : No Translation Path
        g729:8000        To ulaw:8000       : (g729@8000)->(slin@8000)->(ulaw@8000)
        g729:8000        To alaw:8000       : (g729@8000)->(slin@8000)->(alaw@8000)
        g729:8000        To gsm:8000        : (g729@8000)->(slin@8000)->(gsm@8000)
        g729:8000        To g726:8000       : (g729@8000)->(slin@8000)->(g726@8000)
        g729:8000        To g726aal2:8000   : (g729@8000)->(slin@8000)->(g726aal2@8000)
        g729:8000        To adpcm:8000      : (g729@8000)->(slin@8000)->(adpcm@8000)
        g729:8000        To slin:8000       : (g729@8000)->(slin@8000)
        g729:8000        To slin:12000      : (g729@8000)->(slin@8000)->(slin@12000)
        g729:8000        To slin:16000      : (g729@8000)->(slin@8000)->(slin@16000)
        g729:8000        To slin:24000      : (g729@8000)->(slin@8000)->(slin@24000)
        g729:8000        To slin:32000      : (g729@8000)->(slin@8000)->(slin@32000)
        g729:8000        To slin:44100      : (g729@8000)->(slin@8000)->(slin@44100)
        g729:8000        To slin:48000      : (g729@8000)->(slin@8000)->(slin@48000)
        g729:8000        To slin:96000      : (g729@8000)->(slin@8000)->(slin@96000)
        g729:8000        To slin:192000     : (g729@8000)->(slin@8000)->(slin@192000)
        g729:8000        To lpc10:8000      : (g729@8000)->(slin@8000)->(lpc10@8000)
        g729:8000        To speex:8000      : (g729@8000)->(slin@8000)->(speex@8000)
        g729:8000        To speex:16000     : (g729@8000)->(slin@8000)->(slin@16000)->(speex@16000)
        g729:8000        To speex:32000     : (g729@8000)->(slin@8000)->(slin@32000)->(speex@32000)
        g729:8000        To ilbc:8000       : (g729@8000)->(slin@8000)->(ilbc@8000)
        g729:8000        To g722:16000      : (g729@8000)->(slin@8000)->(g722@16000)
        g729:8000        To siren7:16000    : (g729@8000)->(slin@8000)->(slin@16000)->(siren7@16000)
        g729:8000        To siren14:32000   : (g729@8000)->(slin@8000)->(slin@32000)->(siren14@32000)
        g729:8000        To g719:48000      : No Translation Path
        g729:8000        To opus:48000      : (g729@8000)->(slin@8000)->(slin@48000)->(opus@48000)
        g729:8000        To none:8000       : No Translation Path
        g729:8000        To silk:8000       : (g729@8000)->(slin@8000)->(silk@8000)
        g729:8000        To silk:12000      : (g729@8000)->(slin@8000)->(slin@12000)->(silk@12000)
        g729:8000        To silk:16000      : (g729@8000)->(slin@8000)->(slin@16000)->(silk@16000)
        g729:8000        To silk:24000      : (g729@8000)->(slin@8000)->(slin@24000)->(silk@24000)

`

Is the g729 package is installed properly, if not, please install once again.

`
dpkg -l | grep g729
ii  asterisk20-g729                                  2005-1.sng12                              amd64        Asterisk 20 G729 Codec

`

 Also is your system activated with deployment ID ?
  



Thanks

**Comment by guayomejiae** [comment]
Thanks @girishmani1992 
dpkg shows exactly the same info as your message. 

I don't know what happened with the activation that was already done, but when checking with the command "fwconsole sysadmin info" the activation was present but without ID. I performed a HW reset in the portal and activated again successfully.

All I had to do was press the "apply" button for the translation paths to appear for g729 and alaw:

`
--- Translation paths SRC Codec "g729" sample rate 8000 ---
g729:8000        To codec2:8000     : (g729@8000)->(slin@8000)->(codec2@8000)
g729:8000        To g723:8000       : No Translation Path
g729:8000        To ulaw:8000       : (g729@8000)->(slin@8000)->(ulaw@8000)
g729:8000        To alaw:8000       : (g729@8000)->(slin@8000)->(alaw@8000)
g729:8000        To gsm:8000        : (g729@8000)->(slin@8000)->(gsm@8000)
g729:8000        To g726:8000       : (g729@8000)->(slin@8000)->(g726@8000)
g729:8000        To g726aal2:8000   : (g729@8000)->(slin@8000)->(g726aal2@8000)

`

Thanks for the help, I will take care of what caused the deactivation, maybe some hardware change.
